pkg/api/raw.go

```diff
@@ -0,0 +1,107 @@
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/gorilla/mux"
+
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/file/splitter"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+type rawPostResponse struct {
+	Hash swarm.Address `json:"hash"`
+}
+
+// rawUploadHandler handles upload of raw binary data of arbitrary length.
+func (s *server) rawUploadHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	responseObject, err := s.splitUpload(ctx, r.Body, r.ContentLength)
+	if err != nil {
+		jsonhttp.InternalServerError(w, err.Error())
+		return
+	}
+	jsonhttp.OK(w, responseObject)
+}
+
+func (s *server) splitUpload(ctx context.Context, r io.ReadCloser, l int64) (interface{}, error) {
+	chunkPipe := file.NewChunkPipe()
+	go func() {
+		buf := make([]byte, swarm.ChunkSize)
+		c, err := io.CopyBuffer(chunkPipe, r, buf)
+		if err != nil {
+			s.Logger.Debugf("raw: io error %d: %v", c, err)
+			s.Logger.Error("raw: io error")
+			return
+		}
+		if c != l {
+			s.Logger.Debugf("raw: read count mismatch %d: %v", c, err)
+			s.Logger.Error("raw: read count mismatch")
+			return
+		}
+		err = chunkPipe.Close()
+		if err != nil {
+			s.Logger.Errorf("raw: incomplete file write close %v", err)
+			s.Logger.Error("raw: incomplete file write close")
+		}
+	}()
+	sp := splitter.NewSimpleSplitter(s.Storer)
+	address, err := sp.Split(ctx, chunkPipe, l)
+	if err != nil {
+		s.Logger.Debugf("raw: split error %s: %v", address, err)
+		s.Logger.Error("raw: split error")
+		return errors.New("upload error"), nil
+	}
+	return rawPostResponse{Hash: address}, nil
+}
+
+// rawGetHandler handles retrieval of raw binary data of arbitrary length.
+func (s *server) rawGetHandler(w http.ResponseWriter, r *http.Request) {
+	addressHex := mux.Vars(r)["address"]
+	ctx := r.Context()
+
+	address, err := swarm.ParseHexAddress(addressHex)
+	if err != nil {
+		s.Logger.Debugf("raw: parse address %s: %v", addressHex, err)
+		s.Logger.Error("raw: parse address error")
+		jsonhttp.BadRequest(w, "invalid address")
+		return
+	}
+
+	j := joiner.NewSimpleJoiner(s.Storer)
+
+	dataSize, err := j.Size(ctx, address)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			s.Logger.Debugf("raw: not found %s: %v", address, err)
+			s.Logger.Error("raw: not found")
+			jsonhttp.NotFound(w, "not found")
+			return
+		}
+		s.Logger.Debugf("raw: invalid root chunk %s: %v", address, err)
+		s.Logger.Error("raw: invalid root chunk")
+		jsonhttp.BadRequest(w, "invalid root chunk")
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Length", fmt.Sprintf("%d", dataSize))
+	err = file.JoinReadAll(j, address, w)
+	if err != nil {
+		s.Logger.Errorf("raw: data write %s: %v", address, err)
+		s.Logger.Error("raw: data input error")
+		return
```

Comment:
No need to return.

If no bytes are written to `w` (file.JoinReadAll returning an error before writing to writer), the client will receive an empty response. In that case, we should return a json response with internal server error. This can be left for later to be solved as I think that it will prolong this PR.

Additionally, Content-Type and Content-Length should be written only if file.JoinReadAll does not return an error before it writes to the writer.

Reply:
both of your points describe something that means `JoinReadAll` has to change, which would hold up the PR...

Reply:
Actually, JoinReadAll does not have to change. There are multiple ways to implement changes from my comments.

Reply:
so you mean wrap the writer _again_ and see if it's 0 or not?

Reply:
I'll just change readall to return read count

Reply:
> so you mean wrap the writer again and see if it's 0 or not?

This writer is not wrapped at all. Write method is called directly in JoinReadAll. Anyways, if the functionality is broken, why not to wrap.